### PR TITLE
[EDL] Use std::chrono

### PR DESCRIPTION
**Pull request by enen92**
## Description
This PR moves the EDL interface to `std::chrono` so that we can benefit from type safety, expressive syntax and builtin conversions for duration. No functional changes are added.
Runtime tested and tests passed :)

### CODE REVIEW

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 1]
```diff
@@ -49,16 +49,16 @@ TEST_F(TestEdl, TestParsingMplayerTimeBasedEDL)
   // file has only 1 cut starting at 5.3 secs and ending at 7.1 secs
   // so total cut time should be 1.8 seconds (1800 msec)
   EXPECT_EQ(edl.HasCuts(), true);
-  EXPECT_EQ(edl.GetTotalCutTime(), 1.8 * 1000);
+  EXPECT_EQ(edl.GetTotalCutTime(), 1800ms);
```
Comment (fuzzard) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:52]: Should there be an include in this file now?

```
 #include <chrono>
```

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -1281,22 +1281,24 @@ void CVideoPlayer::Prepare()
    * if there was a start time specified as part of the "Start from where last stopped" (aka
    * auto-resume) feature or if there is an EDL cut or commercial break that starts at time 0.
    */
-  int starttime = 0;
+  std::chrono::milliseconds starttime = 0ms;
```
Comment (fuzzard) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:1284]: An include here as well?

File: xbmc/guilib/guiinfo/PlayerGUIInfo.cpp [hunk 3]
```diff
@@ -672,8 +672,8 @@ std::vector<std::pair<float, float>> CPlayerGUIInfo::GetEditList(const CDataCach
   const std::vector<EDL::Edit>& edits = data.GetEditList();
   for (const auto& edit : edits)
   {
-    float editStart = edit.start * 100.0f / duration;
-    float editEnd = edit.end * 100.0f / duration;
+    float editStart = edit.start.count() * 100.0f / duration;
```
Comment (fuzzard) [inline xbmc/guilib/guiinfo/PlayerGUIInfo.cpp:675]: Same again re include

File: xbmc/pvr/PVREdl.cpp [hunk 4]
```diff
@@ -38,8 +38,8 @@ std::vector<EDL::Edit> CPVREdl::GetEdits(const CFileItem& item)
   for (const auto& entry : edl)
   {
     EDL::Edit edit;
-    edit.start = entry.start;
-    edit.end = entry.end;
+    edit.start = std::chrono::milliseconds(entry.start);
```
Comment (fuzzard) [inline xbmc/pvr/PVREdl.cpp:41]: include

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp (hunk 1 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:52]: The Edit struct from the EDLEdit.h header is based on chrono and includes it, but I'll follow your suggestion in case it is somehow forward declared in the future

**Comment by enen92** [comment]
jenkins build this please

File: xbmc/cores/EdlEdit.h [hunk 5]
```diff
@@ -8,6 +8,10 @@
 
 #pragma once
 
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Comment (garbear) [inline xbmc/cores/EdlEdit.h:13]: Shouldn't use a `using` statement in a header.

Comment (garbear) [reply xbmc/cores/EdlEdit.h:13]: Can you brace initialize the timing members?

Comment (enen92) [reply xbmc/cores/EdlEdit.h]: That's something I did not think about before but the reason is definitely on your side. Thanks!

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 6]
```diff
@@ -2411,9 +2418,9 @@ void CVideoPlayer::CheckAutoSceneSkip()
       m_CurrentVideo.inited == false)
     return;
 
-  const int64_t clock = GetTime();
+  const std::chrono::milliseconds clock = std::chrono::milliseconds(GetTime());
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:2421]: Isn’t the compiler able to convert GetTime return value to the target type?

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 7]
```diff
@@ -2700,7 +2708,9 @@ void CVideoPlayer::HandleMessages()
       if (msg.GetRelative())
         time = (m_clock.GetClock() + m_State.time_offset) / 1000l + time;
 
-      time = msg.GetRestore() ? m_Edl.GetTimeAfterRestoringCuts(time) : time;
+      time = msg.GetRestore()
+                 ? m_Edl.GetTimeAfterRestoringCuts(std::chrono::milliseconds(lround(time))).count()
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:2712]: Not directly related, but std::lround?

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 8]
```diff
@@ -49,16 +51,16 @@ TEST_F(TestEdl, TestParsingMplayerTimeBasedEDL)
...
   EXPECT_EQ(edl.GetCutMarkers().size(), 1);
-  EXPECT_EQ(edl.GetCutMarkers().at(0), 5.3 * 1000); // 5.3 secs
+  EXPECT_EQ(edl.GetCutMarkers().at(0), 5300ms); // 5.3 secs
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:56]: Remove the comment? Code is self-explaining now.

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 9]
```diff
@@ -126,9 +132,9 @@ TEST_F(TestEdl, TestParsingMplayerTimeBasedInterleavedCutsEDL)
   EXPECT_EQ(edl.HasEdits(), true);
   EXPECT_EQ(edl.GetCutMarkers().size(), 2);
   // lets check the total cut time matches the sum of the two cut durations defined in the file
-  EXPECT_EQ(edl.GetTotalCutTime(), ((7.1 - 5.3) + (19 - 18)) * 1000);
+  EXPECT_EQ(edl.GetTotalCutTime(), 2800ms);
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:135]: Would it make sense to keep the math explaining where the 2.8 secs come from?

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 10]
```diff
@@ -274,23 +288,27 @@ TEST_F(TestEdl, TestComSkipVersion2EDL)
   EXPECT_EQ(found, true);
   // fps is obtained from the file as it always takes precedence (note we supplied 0 above),
   // the EDL file has the value of 2500 for fps. kodi converts this to 25 fps by dividing by a factor of 100
-  const float fpsInEdlFile = 2500 / 100;
+  const double fpsInEdlFile = 2500 / 100;
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:291]: Why this change?

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 11]
```diff
@@ -325,17 +343,17 @@ TEST_F(TestEdl, TestCommBreakAdvancedSettings)
...
+  EXPECT_EQ(edl.GetEditList().at(0).start, 10s);
+  EXPECT_EQ(edl.GetEditList().at(0).end, 22s);
+  EXPECT_EQ(edl.GetEditList().at(1).start, 30s);
+  EXPECT_EQ(edl.GetEditList().at(1).end, 32s);
+  EXPECT_EQ(edl.GetEditList().at(2).start, 37s);
+  EXPECT_EQ(edl.GetEditList().at(2).end, 50s);
+  EXPECT_EQ(edl.GetEditList().at(3).start, 52s);
+  EXPECT_EQ(edl.GetEditList().at(3).end, 60s);
+  EXPECT_EQ(edl.GetEditList().at(4).start, 62s);
+  EXPECT_EQ(edl.GetEditList().at(4).end, 65100ms);
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:355]: 65.1s , for consistency with the rest of the code block?

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp [hunk 12]
```diff
@@ -458,8 +476,7 @@ TEST_F(TestEdl, TestMergeSmallCommbreaksAdvanced)
   // kodi should merge all commbreaks into two
   EXPECT_EQ(edl.GetEditList().size(), 2);
   // second edit of the original file + third one
-  EXPECT_EQ(edl.GetEditList().at(0).end - edl.GetEditList().at(0).start, (32 - 10) * 1000);
+  EXPECT_EQ(edl.GetEditList().at(0).end - edl.GetEditList().at(0).start, 32s - 10s);
   // 4th, 5th and 6th commbreaks joined
-  EXPECT_EQ(edl.GetEditList().at(1).end - edl.GetEditList().at(1).start,
-            std::lround((65.1 - 37) * 1000));
+  EXPECT_EQ(edl.GetEditList().at(1).end - edl.GetEditList().at(1).start, 65100ms - 37s);
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:481]: 65.1s ?

**Review by ksooo — COMMENTED** [review]
Nice. Only some minors / questions.

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp (hunk 10 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:291]: Because it could lead to wrong approximations (floating point) when doing the later comparisons against seconds with decimal places (the internal value is correct but the conversion from tests in seconds can lead the test to fail in the CI). We internally store milliseconds and all the comparisons are done in milliseconds. Some tests are written comparing to seconds just for increased readability.

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp (hunk 11 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:355]: Yeah it was set like that before but it leads to errors (floating point) in our Jenkins workers. Works fine on my machines though. Hence I opted to keep this one as ms

File: xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp (hunk 12 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/test/edl/TestEdl.cpp:481]: Same issue as above, avoiding the float on the seconds value

**Review by ksooo — APPROVED** [review]
LGTM

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 13]
```diff
@@ -179,9 +180,9 @@ bool CEdl::ReadEdl(const std::string& mediaFilePath, float fps)
           {
             fieldParts[1] = fieldParts[1].substr(0, 3);
           }
-          editStartEnd[i] =
-              static_cast<int64_t>(StringUtils::TimeStringToSeconds(fieldParts[0])) * 1000 +
-              std::atoi(fieldParts[1].c_str()); // seconds to ms
+          editStartEnd[i] = std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::seconds(StringUtils::TimeStringToSeconds(fieldParts[0])));
+          editStartEnd[i] += std::chrono::milliseconds(std::atoi(fieldParts[1].c_str()));
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/Edl.cpp:185]: Could use `std::stoi` to avoid the `c_str` call but may have to handle the possible exception.

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 14]
```diff
@@ -658,20 +669,26 @@ bool CEdl::AddEdit(const Edit& newEdit)
      * the start (autowait) and automatically rewind by a bit (autowind) at the end of the commercial
      * break.
      */
-    int autowait = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_iEdlCommBreakAutowait * 1000; // seconds -> ms
-    int autowind = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_iEdlCommBreakAutowind * 1000; // seconds -> ms
-
-    if (edit.start > 0) // Only autowait if not at the start.
+    std::chrono::milliseconds autowait = std::chrono::duration_cast<std::chrono::milliseconds>(
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/Edl.cpp:672]: Could use `auto` here

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 15]
```diff
@@ -3268,19 +3279,19 @@ bool CVideoPlayer::SeekScene(Direction seekDirection)
    * There is a 5 second grace period applied when seeking for scenes backwards. If there is no
    * grace period applied it is impossible to go backwards past a scene marker.
    */
-  int64_t clock = GetTime();
-  if (seekDirection == Direction::BACKWARD && clock > 5 * 1000) // 5 seconds
-    clock -= 5 * 1000;
+  std::chrono::milliseconds clock = std::chrono::milliseconds(GetTime());
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:3282]:
```suggestion
  auto clock = std::chrono::milliseconds(GetTime());
```

File: xbmc/cores/VideoPlayer/Edl.cpp (hunk 13 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/Edl.cpp]: ended up using `std::from_chars` since it doesn't throw and leave the resulting value unmodified in case of errors. Seem to fit the implementation here better as we're appending to the already existing value
